druid-shell/src/backend/x11/dialog.rs

```diff
@@ -0,0 +1,169 @@
+) -> FileDialogToken {
+    dialog(window, idle, options, true)
+}
+
+pub(crate) fn save_file(
+    window: u32,
+    idle: IdleHandle,
+    options: FileDialogOptions,
+) -> FileDialogToken {
+    dialog(window, idle, options, false)
+}
+
+fn dialog(
+    window: u32,
+    idle: IdleHandle,
+    mut options: FileDialogOptions,
+    open: bool,
+) -> FileDialogToken {
+    let tok = FileDialogToken::next();
+
+    std::thread::spawn(move || {
+        if let Err(e) = block_on(async {
+            let mut handle = XcbHandle::empty();
+            handle.window = window;
+
+            let conn = zbus::Connection::session().await?;
+            let proxy = file_chooser::FileChooserProxy::new(&conn).await?;
+            let id = WindowIdentifier::from_raw_handle(&RawWindowHandle::Xcb(handle));
+            let multi = options.multi_selection;
+
+            let title_owned = options.title.take();
+            let title = match (open, options.select_directories) {
+                (true, true) => "Open Folder",
+                (true, false) => "Open File",
+                (false, _) => "Save File",
+            };
+            let title = title_owned.as_deref().unwrap_or(title);
+
+            let callback = move |uris: &[String]| {
+                let mut paths = uris.iter().filter_map(|s| {
+                    s.strip_prefix("file://").or_else(|| {
+                        warn!("expected path '{}' to start with 'file://'", s);
+                        None
+                    })
+                });
+                if multi && open {
+                    let infos = paths
+                        .map(|p| FileInfo {
+                            path: p.into(),
+                            format: None,
+                        })
+                        .collect();
+                    idle.add_idle_callback(move |handler| handler.open_files(tok, infos));
+                } else if !multi {
+                    if uris.len() > 2 {
+                        warn!(
+                            "expected one path (got {}), returning only the first",
+                            uris.len()
+                        );
+                    }
+                    let info = paths.next().map(|p| FileInfo {
+                        path: p.into(),
+                        format: None,
+                    });
+                    if open {
+                        idle.add_idle_callback(move |handler| handler.open_file(tok, info));
+                    } else {
+                        idle.add_idle_callback(move |handler| handler.save_as(tok, info));
+                    }
+                } else {
+                    warn!("cannot save multiple paths");
+                }
+            };
+
+            if open {
+                callback(proxy.open_file(&id, title, options.into()).await?.uris());
+            } else {
+                callback(proxy.save_file(&id, title, options.into()).await?.uris());
+            };
```

Comment:
any reason to use a callback? can't we just `let uris = if open ..... ` and then inline callback

Reply:
Yeah, the result of `uris()` is borrowed from the objects returned from `open_file` or `save_file`, so that version will drop the temporary objects too soon. But it works with some extra temporary variables, and indeed I think it's nicer. Thanks!